File: frontend/test/shared/utils/jwt_decode_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:weekplanner/shared/utils/jwt_decode.dart';

String _createJwt(Map<String, dynamic> payload) {
  final header = base64Url.encode(utf8.encode('{"alg":"HS256","typ":"JWT"}'));
  final body = base64Url.encode(utf8.encode(json.encode(payload)));
  return '$header.$body.fake_signature';
}

void main() {
  group('JwtDecode', () {
    test('getUserId extracts user_id claim', () {
      final token = _createJwt({'user_id': '42'});
      expect(JwtDecode.getUserId(token), '42');
    });

    test('getUserId returns empty string when claim is missing', () {
      final token = _createJwt({'sub': 'something'});
      expect(JwtDecode.getUserId(token), '');
    });

    test('getOrgRoles extracts org_roles map', () {
      final token = _createJwt({
        'org_roles': {'1': 'admin', '2': 'member'},
      });
      final roles = JwtDecode.getOrgRoles(token);
      expect(roles, {'1': 'admin', '2': 'member'});
    });

    test('getOrgRoles returns empty map when claim is missing', () {
      final token = _createJwt({});
      expect(JwtDecode.getOrgRoles(token), isEmpty);
    });

    test('isExpired returns true when token is expired', () {
      final pastExp = DateTime.now().subtract(const Duration(hours: 1));
      final token = _createJwt({
        'exp': pastExp.millisecondsSinceEpoch ~/ 1000,
      });
      expect(JwtDecode.isExpired(token), true);
    });

    test('isExpired returns false when token is still valid', () {
      final futureExp = DateTime.now().add(const Duration(hours: 1));
      final token = _createJwt({
        'exp': futureExp.millisecondsSinceEpoch ~/ 1000,
      });
      expect(JwtDecode.isExpired(token), false);
    });

    test('isExpired returns true when exp claim is missing', () {
      final token = _createJwt({});
      expect(JwtDecode.isExpired(token), true);
    });

    test('throws FormatException for invalid token', () {
      expect(() => JwtDecode.getUserId('not.a.valid'), throwsFormatException);
    });

    test('handles numeric user_id', () {
      final token = _createJwt({'user_id': 123});
      expect(JwtDecode.getUserId(token), '123');
    });
  });
}

File: frontend/__tests__/jwtDecode.test.ts
import { isTokenExpired, getUserIdFromToken, getOrgRolesFromToken } from "../utils/jwtDecode";

describe("isTokenExpired", () => {
  it("returns true for expired token", () => {
    const expiredToken = "header.expiredPayload.signature";
    expect(isTokenExpired(expiredToken)).toBe(true);
  });

  it("returns false for valid token", () => {
    const validToken = "header." + btoa(JSON.stringify({ exp: Date.now() / 1000 + 1000 })) + ".signature";
    expect(isTokenExpired(validToken)).toBe(false);
  });
});

describe("getUserIdFromToken", () => {
  it("extracts user ID from token sub claim", () => {
    const userId = "user-123";
    const payload = { user_id: userId, exp: Date.now() / 1000 + 3600 };
    const token = "header." + btoa(JSON.stringify(payload)) + ".signature";

    expect(getUserIdFromToken(token)).toBe(userId);
  });

  it("handles malformed token gracefully", () => {
    const malformedToken = "not.a.valid.token";
    // Should not throw, returns empty string for invalid tokens
    const result = getUserIdFromToken(malformedToken);
    expect(result).toBe("");
  });
});

describe("getOrgRolesFromToken (Core Auth)", () => {
  it("extracts org_roles from Core JWT", () => {
    const orgRoles = { "1": "owner", "5": "member", "10": "admin" };
    const payload = {
      sub: "user-123",
      exp: Date.now() / 1000 + 3600,
      org_roles: orgRoles
    };
    const token = "header." + btoa(JSON.stringify(payload)) + ".signature";

    expect(getOrgRolesFromToken(token)).toEqual(orgRoles);
  });

  it("returns empty object when org_roles is missing", () => {
    const payload = { sub: "user-123", exp: Date.now() / 1000 + 3600 };
    const token = "header." + btoa(JSON.stringify(payload)) + ".signature";

    expect(getOrgRolesFromToken(token)).toEqual({});
  });

  it("returns empty object for malformed token", () => {
    const malformedToken = "not.a.valid.token";

    expect(getOrgRolesFromToken(malformedToken)).toEqual({});
  });

  it("handles empty org_roles object", () => {
    const payload = {
      sub: "user-123",
      exp: Date.now() / 1000 + 3600,
      org_roles: {}
    };
    const token = "header." + btoa(JSON.stringify(payload)) + ".signature";

    expect(getOrgRolesFromToken(token)).toEqual({});
  });

  it("preserves org_roles structure from Core", () => {
    const orgRoles = {
      "1": "owner",
      "2": "admin",
      "3": "member",
      "100": "member",
      "999": "admin"
    };
    const payload = {
      sub: "user-456",
      exp: Date.now() / 1000 + 3600,
      org_roles: orgRoles
    };
    const token = "header." + btoa(JSON.stringify(payload)) + ".signature";

    const result = getOrgRolesFromToken(token);
    expect(result).toEqual(orgRoles);
    expect(Object.keys(result)).toHaveLength(5);
    expect(result["1"]).toBe("owner");
    expect(result["100"]).toBe("member");
  });
});
